---
title: CA1064：异常应该是公共的（代码分析）
description: 了解代码分析规则 CA1064：异常应该是公共的
ms.date: 11/04/2016
ms.topic: reference
f1_keywords:
- CA1064
- ExceptionsShouldBePublic
helpviewer_keywords:
- ExceptionsShouldBePublic
- CA1064
author: gewarren
ms.author: gewarren
ms.openlocfilehash: 306085feffd6b61b75c268dd1719d037a65a0a8a
ms.sourcegitcommit: 05d0087dfca85aac9ca2960f86c5efd218bf833f
ms.translationtype: HT
ms.contentlocale: zh-CN
ms.lasthandoff: 03/30/2021
ms.locfileid: "99546577"
---
# <a name="ca1064-exceptions-should-be-public"></a><span data-ttu-id="7e32f-103">CA1064:异常应该是公共的</span><span class="sxs-lookup"><span data-stu-id="7e32f-103">CA1064: Exceptions should be public</span></span>

| | <span data-ttu-id="7e32f-104">值</span><span class="sxs-lookup"><span data-stu-id="7e32f-104">Value</span></span> |
|-|-|
| <span data-ttu-id="7e32f-105">**规则 ID**</span><span class="sxs-lookup"><span data-stu-id="7e32f-105">**Rule ID**</span></span> |<span data-ttu-id="7e32f-106">CA1064</span><span class="sxs-lookup"><span data-stu-id="7e32f-106">CA1064</span></span>|
| <span data-ttu-id="7e32f-107">**类别**</span><span class="sxs-lookup"><span data-stu-id="7e32f-107">**Category**</span></span> |[<span data-ttu-id="7e32f-108">Microsoft.Design</span><span class="sxs-lookup"><span data-stu-id="7e32f-108">Microsoft.Design</span></span>](design-warnings.md)|
| <span data-ttu-id="7e32f-109">修复是中断修复还是非中断修复</span><span class="sxs-lookup"><span data-stu-id="7e32f-109">**Fix is breaking or non-breaking**</span></span> |<span data-ttu-id="7e32f-110">非中断</span><span class="sxs-lookup"><span data-stu-id="7e32f-110">Non-breaking</span></span>|

## <a name="cause"></a><span data-ttu-id="7e32f-111">原因</span><span class="sxs-lookup"><span data-stu-id="7e32f-111">Cause</span></span>

<span data-ttu-id="7e32f-112">非公共异常直接派生自 <xref:System.Exception>、<xref:System.SystemException> 或 <xref:System.ApplicationException>。</span><span class="sxs-lookup"><span data-stu-id="7e32f-112">A non-public exception derives directly from <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>.</span></span>

## <a name="rule-description"></a><span data-ttu-id="7e32f-113">规则说明</span><span class="sxs-lookup"><span data-stu-id="7e32f-113">Rule description</span></span>

<span data-ttu-id="7e32f-114">内部异常仅在其自己的内部范围内可见。</span><span class="sxs-lookup"><span data-stu-id="7e32f-114">An internal exception is only visible inside its own internal scope.</span></span> <span data-ttu-id="7e32f-115">当异常超出内部范围后，只能使用基异常来捕获该异常。</span><span class="sxs-lookup"><span data-stu-id="7e32f-115">After the exception falls outside the internal scope, only the base exception can be used to catch the exception.</span></span> <span data-ttu-id="7e32f-116">如果内部异常继承自 <xref:System.Exception>、<xref:System.SystemException> 或 <xref:System.ApplicationException>，则外部代码将没有足够的信息来了解如何处理该异常。</span><span class="sxs-lookup"><span data-stu-id="7e32f-116">If the internal exception is inherited from <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>, the external code will not have sufficient information to know what to do with the exception.</span></span>

<span data-ttu-id="7e32f-117">但是，如果代码有一个公共异常，稍后会用作内部异常的基异常，则有理由认为后续代码将能够对该基异常进行智能化操作。</span><span class="sxs-lookup"><span data-stu-id="7e32f-117">But, if the code has a public exception that later is used as the base for a internal exception, it is reasonable to assume the code further out will be able to do something intelligent with the base exception.</span></span> <span data-ttu-id="7e32f-118">该公共异常将会比 <xref:System.Exception>、<xref:System.SystemException> 或 <xref:System.ApplicationException> 提供更多的信息。</span><span class="sxs-lookup"><span data-stu-id="7e32f-118">The public exception will have more information than what is provided by <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>.</span></span>

## <a name="how-to-fix-violations"></a><span data-ttu-id="7e32f-119">如何解决冲突</span><span class="sxs-lookup"><span data-stu-id="7e32f-119">How to fix violations</span></span>

<span data-ttu-id="7e32f-120">使异常成为公共异常，或从不是 <xref:System.Exception>、<xref:System.SystemException> 或 <xref:System.ApplicationException> 的公共异常派生内部异常。</span><span class="sxs-lookup"><span data-stu-id="7e32f-120">Make the exception public, or derive the internal exception from a public exception that is not <xref:System.Exception>, <xref:System.SystemException>, or <xref:System.ApplicationException>.</span></span>

## <a name="when-to-suppress-warnings"></a><span data-ttu-id="7e32f-121">何时禁止显示警告</span><span class="sxs-lookup"><span data-stu-id="7e32f-121">When to suppress warnings</span></span>

<span data-ttu-id="7e32f-122">如果确定在所有情况下私有异常都将在其自己的内部范围内被捕获，则禁止显示此规则的消息。</span><span class="sxs-lookup"><span data-stu-id="7e32f-122">Suppress a message from this rule if you are sure in all cases that the private exception will be caught within its own internal scope.</span></span>

## <a name="example"></a><span data-ttu-id="7e32f-123">示例</span><span class="sxs-lookup"><span data-stu-id="7e32f-123">Example</span></span>

<span data-ttu-id="7e32f-124">此规则在第一个示例方法 FirstCustomException 上触发，因为 exception 类直接派生自 Exception ，并且是内部类。</span><span class="sxs-lookup"><span data-stu-id="7e32f-124">This rule fires on the first example method, FirstCustomException because the exception class derives directly from Exception and is internal.</span></span> <span data-ttu-id="7e32f-125">此规则不会在 SecondCustomException 类上触发，尽管该类也直接派生自 Exception，但该类声明为公共类。</span><span class="sxs-lookup"><span data-stu-id="7e32f-125">The rule does not fire on the SecondCustomException class because although the class also derives directly from Exception, the class is declared public.</span></span> <span data-ttu-id="7e32f-126">第三个类也不会触发该规则，因为它并非直接派生自 <xref:System.Exception?displayProperty=fullName>、<xref:System.SystemException?displayProperty=fullName> 或 <xref:System.ApplicationException?displayProperty=fullName>。</span><span class="sxs-lookup"><span data-stu-id="7e32f-126">The third class also does not fire the rule because it does not derive directly from <xref:System.Exception?displayProperty=fullName>, <xref:System.SystemException?displayProperty=fullName>, or <xref:System.ApplicationException?displayProperty=fullName>.</span></span>

:::code language="csharp" source="snippets/csharp/all-rules/ca1064.cs" id="snippet1":::
